Sources/System/SystemString.swift

```diff
@@ -0,0 +1,289 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+// A platform-native character representation, currently used for file paths
+internal struct SystemChar: RawRepresentable, Comparable, Hashable, Codable {
+  internal typealias RawValue = CInterop.PlatformChar
+
+  internal var rawValue: RawValue
+
+  internal init(rawValue: RawValue) { self.rawValue = rawValue }
+
+  internal init(_ rawValue: RawValue) { self.init(rawValue: rawValue) }
+
+  static func < (lhs: SystemChar, rhs: SystemChar) -> Bool {
+    lhs.rawValue < rhs.rawValue
+  }
+}
+
+extension SystemChar {
+  internal init(ascii: Unicode.Scalar) {
+    self.init(rawValue: numericCast(UInt8(ascii: ascii)))
+  }
+  internal init(codeUnit: CInterop.PlatformUnicodeEncoding.CodeUnit) {
+    self.init(rawValue: codeUnit._platformChar)
+  }
+
+  internal static var null: SystemChar { SystemChar(0x0) }
+  internal static var slash: SystemChar { SystemChar(ascii: "/") }
+  internal static var backslash: SystemChar { SystemChar(ascii: #"\"#) }
+  internal static var dot: SystemChar { SystemChar(ascii: ".") }
+  internal static var colon: SystemChar { SystemChar(ascii: ":") }
+  internal static var question: SystemChar { SystemChar(ascii: "?") }
+
+  internal var codeUnit: CInterop.PlatformUnicodeEncoding.CodeUnit {
+    rawValue._platformCodeUnit
+  }
+
+  internal var asciiScalar: Unicode.Scalar? {
+    guard isASCII else { return nil }
+    return Unicode.Scalar(UInt8(truncatingIfNeeded: rawValue))
+  }
+
+  internal var isASCII: Bool {
+    (0...0x7F).contains(rawValue)
+  }
+
+  internal var isLetter: Bool {
+    guard isASCII else { return false }
+    let asciiRaw: UInt8 = numericCast(rawValue)
+    return (UInt8(ascii: "a") ..< UInt8(ascii: "z")).contains(asciiRaw) ||
+           (UInt8(ascii: "A") ..< UInt8(ascii: "Z")).contains(asciiRaw)
+  }
+}
+
+// A platform-native string representation, currently for file paths
+//
+// Always null-terminated.
+internal struct SystemString {
+  internal typealias Storage = [SystemChar]
+  internal var nullTerminatedStorage: Storage
+}
+
+extension SystemString {
+  internal init() {
+    self.nullTerminatedStorage = [.null]
```

Comment:
It can be, but IIRC how you declare a global matters and we don't want to unnecessarily dirty memory. I never understood all the differences (especially as they interact with `@usableFromInline`, which wouldn't be relevant here yet).